### PR TITLE
BUG: Fix Series nunique groupby with object dtype

### LINKED ISSUE #11077: Broken nunique on Series group by
The following code works in 0.16.2 and not in latest master:

```python
data = pd.DataFrame(
    [[100, 1, 'Alice'],
     [200, 2, 'Bob'],
     [300, 3, 'Charlie'],
     [-400, 4, 'Dan'],
     [500, 5, 'Edith']],
    columns=['amount', 'id', 'name']
)

expected = data.groupby(['id', 'amount'])['name'].nunique()
```

Going to bisect this today unless someone beats me to it.

### PR DESCRIPTION
**Pull request by cpcloud**
closes #11077

### CODE REVIEW

**Comment by jreback** [comment]
cc @behzadnouri 

or you want to use do a factorize?

**Comment by behzadnouri** [comment]
```diff
diff --git a/pandas/core/groupby.py b/pandas/core/groupby.py
index f34fd6e..39be706 100644
--- a/pandas/core/groupby.py
+++ b/pandas/core/groupby.py
@@ -2565,13 +2565,21 @@ class SeriesGroupBy(GroupBy):
         ids, _, _ = self.grouper.group_info
         val = self.obj.get_values()
 
-        sorter = np.lexsort((val, ids))
+        try:
+            sorter = np.lexsort((val, ids))
+        except TypeError:
+            val, _ = algos.factorize(val, sort=False)
+            sorter = np.lexsort((val, ids))
+            isnull = lambda a: a == -1
+        else:
+            isnull = com.isnull
+
         ids, val = ids[sorter], val[sorter]
 
         # group boundries are where group ids change
         # unique observations are where sorted values change
-        idx = com._ensure_int64(np.r_[0, 1 + np.nonzero(ids[1:] != ids[:-1])[0]])
-        inc = com._ensure_int64(np.r_[1, val[1:] != val[:-1]])
+        idx = np.r_[0, 1 + np.nonzero(ids[1:] != ids[:-1])[0]]
+        inc = np.r_[1, val[1:] != val[:-1]]
 
         # 1st item of each group is a new unique observation
         mask = isnull(val)
```

**Comment by cpcloud** [comment]
@behzadnouri If you submit that as a PR to my fork, I'll happily merge it in

**Comment by behzadnouri** [comment]
the error says `TypeError: merge sort not available for item 0` even though numpy _does_ have merge sort for objects (both `np.argsort` and `np.sort`). so it may be a bug on numpy side. so, it is better to implement try catch in case future numpy releases do implement lexsort for this case

once, values are factorized, original values are not needed and it is more efficient to work with integer factors, as long as `isnull` function is adjusted accordingly.

**Comment by cpcloud** [comment]
@jreback getting a somewhat strange travis failure here that seems unrelated to my change: https://travis-ci.org/cpcloud/pandas/jobs/80036566

**Comment by cpcloud** [comment]
any ideas what that might be?

**Comment by jreback** [comment]
hmm someone else reported that as well
try restarting it and see

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -2565,13 +2565,21 @@ def nunique(self, dropna=True):
         ids, _, _ = self.grouper.group_info
         val = self.obj.get_values()
 
-        sorter = np.lexsort((val, ids))
+        try:
+            sorter = np.lexsort((val, ids))
+        except TypeError:
+            val, _ = algos.factorize(val, sort=False)
```
Comment (shoyer) [inline pandas/core/groupby.py]: Note that this means object dtype? Maybe add an assert?

**Comment by cpcloud** [comment]
this is passing on my travis ci fork

File: pandas/core/groupby.py [hunk 2]
```diff
@@ -2565,13 +2565,23 @@ def nunique(self, dropna=True):
         ids, _, _ = self.grouper.group_info
         val = self.obj.get_values()
 
-        sorter = np.lexsort((val, ids))
+        try:
+            sorter = np.lexsort((val, ids))
+        except TypeError:
+            assert val.dtype == object, \
```
Comment (jreback) [inline pandas/core/groupby.py]: add a comment here that this catches ``object`` dtypes

File: pandas/tests/test_groupby.py [hunk 3]
```diff
@@ -5511,6 +5511,21 @@ def test_sort(x):
 
                 g.apply(test_sort)
 
+    def test_nunique_with_object(self):
+        data = pd.DataFrame(
```
Comment (jreback) [inline pandas/tests/test_groupby.py]: GH number as a comment

**Comment by jreback** [comment]
couple comments, squash and merge ok

File: doc/source/whatsnew/v0.17.0.txt [hunk 4]
```diff
@@ -1135,3 +1135,4 @@ Bug Fixes
 - Bug in ``DatetimeIndex`` cannot infer negative freq (:issue:`11018`)
 - Remove use of some deprecated numpy comparison operations, mainly in tests. (:issue:`10569`)
 - Bug in ``Index`` dtype may not applied properly (:issue:`11017`)
+- Bug in Series groupby when calling nunique on an object dtype (:issue:`11077`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: move this to where the other nunique comment is (just put in in the list)

File: pandas/core/groupby.py [hunk 5]
```diff
@@ -2565,13 +2565,23 @@ def nunique(self, dropna=True):
...
+            isnull = lambda a: a == -1
+        else:
+            isnull = com.isnull
+
         ids, val = ids[sorter], val[sorter]
 
         # group boundries are where group ids change
         # unique observations are where sorted values change
-        idx = com._ensure_int64(np.r_[0, 1 + np.nonzero(ids[1:] != ids[:-1])[0]])
-        inc = com._ensure_int64(np.r_[1, val[1:] != val[:-1]])
```
Comment (jreback) [inline pandas/core/groupby.py]: revert these 2 lines, these don't pass on windows ATM. this can possibly be changed, but on another PR/issue

**Comment by jreback** [comment]
@cpcloud can you update (e.g. revert those 2 lines). ok to merge on green after that.

**Comment by jreback** [comment]
ty sir!

**Pull request merged**